src/foam/dao/MongoDAO.java

```diff
@@ -0,0 +1,176 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.dao;
+
+import java.util.List;
+import java.util.Arrays;
+import java.util.Iterator;
+
+import foam.core.FObject;
+import foam.core.X;
+import foam.core.PropertyInfo;
+import foam.mlang.order.Comparator;
+import foam.mlang.predicate.Predicate;
+
+// FObject JSON Parsing
+import foam.core.EmptyX;
+import foam.lib.json.JSONParser;
+import foam.lib.json.Outputter;
+
+// MongoDB Driver Dependencies
+import com.mongodb.MongoClient;
+import com.mongodb.MongoClientURI;
+import com.mongodb.ServerAddress;
+import com.mongodb.MongoCredential;
+
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoCursor;
+
+import org.bson.Document;
+import org.bson.json.JsonWriterSettings;
+import org.bson.json.JsonMode;
+
+
+public class MongoDAO
+        extends AbstractDAO
+{
+  MongoDatabase database;
+
+  private static final int MONGO_OBJECT_PREFIX_LENGTH = 49;
+
+  public MongoDAO(String host, int port, String dbName, String username, String password) {
+    if ( dbName == null || dbName.isEmpty() ) {
+      throw new IllegalArgumentException("Illegal arguments");
+    }
+
+    host = ( host != null ) ? host : "localhost";
+
+    MongoClient mongoClient;
+
+    if ( isUserPassProvided(username, password) ) {
+      MongoCredential credential = MongoCredential.createCredential(username, dbName, password.toCharArray());
+      mongoClient = new MongoClient(new ServerAddress(host, port), Arrays.asList(credential));
+    } else {
+      mongoClient = new MongoClient(host, port);
+    }
+
+    this.database = mongoClient.getDatabase(dbName);
+  }
+
+  public MongoDAO(String dbName, String username, String password) {
+    this("localhost", 27017, dbName, username, password);
+  }
+
+  public MongoDAO(String dbName) {
+    this(dbName, null, null);
+  }
+
+  private boolean isUserPassProvided(String username, String password) {
+    return ( ( username != null ) && ( ! username.isEmpty() ) ) &&
+            ( ( password != null ) && ( ! password.isEmpty() ) );
+  }
+
+  @Override
+  public Sink select_(X x, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
+    if ( sink == null ) {
+      sink = new ListSink();
+    }
+
+    Sink         decorated = decorateSink_(sink, skip, limit, order, predicate);
+    Subscription sub       = new Subscription();
+
+    String collectionName = (String) x.get("collectionName");
+    String collectionClass = (String) x.get("collectionClass");
+    List props = getFObjectProperties(collectionClass, x);
+
+    if ( collectionName == null || collectionName.isEmpty() ) {
```

Comment:
That would be of some use, but as a result the MongoDAO would behave differently than other DAOs.  It would be better to be consistent so someone looking to use a MongoDAO doesn't have to learn a different way of accessing using the DAO.